python-package/lightgbm/callback.py

```diff
@@ -350,12 +353,17 @@ def __call__(self, env: CallbackEnv) -> None:
             self._init(env)
         if not self.enabled:
             return
+        # self.best_score_list is initialized to an empty list
+        first_time_updating_best_score_list = (self.best_score_list == [])
         for i in range(len(env.evaluation_result_list)):
             score = env.evaluation_result_list[i][2]
-            if self.best_score_list[i] is None or self.cmp_op[i](score, self.best_score[i]):
+            if first_time_updating_best_score_list or self.cmp_op[i](score, self.best_score[i]):
                 self.best_score[i] = score
                 self.best_iter[i] = env.iteration
-                self.best_score_list[i] = env.evaluation_result_list
+                if first_time_updating_best_score_list:
+                    self.best_score_list.append(env.evaluation_result_list)
+                else:
+                    self.best_score_list[i] = env.evaluation_result_list
```

Comment:
On `master`, here's how `_EarlyStoppingCallback` currently populates `best_score_list`.

The first time the callback is called, run this `_EarlyStoppingCallback._init()` method.

https://github.com/microsoft/LightGBM/blob/2fe2bf0675ef127a29f9bc49511fc0099cf3e140/python-package/lightgbm/callback.py#L348-L350

In that `_init()` method, ensure that `self.best_score_list` is the same length as `CallbackEnv.evaluation_result_list` by padding it with `None`s.

https://github.com/microsoft/LightGBM/blob/2fe2bf0675ef127a29f9bc49511fc0099cf3e140/python-package/lightgbm/callback.py#L328-L330

And then whenever the callback is called, iterate over `CallbackEnv.evaluation_result_list` and put the corresponding item from it into `self.best_score_list` if:

* it's the very first time the callback was called (`self.best_score_list[i] is None`)
* the new evaluation result is better than the previous one by at least some margin `delta`, which defaults to 0.0 (`self.cmp_op[i](score, self.best_score[i])`

https://github.com/microsoft/LightGBM/blob/2fe2bf0675ef127a29f9bc49511fc0099cf3e140/python-package/lightgbm/callback.py#L355-LL358

So those `None` entries in `self.best_score_list` aren't expected to be there when the first call of `_EarlyStoppingCallback` completes. They're not visible to user code, and nothing else in `lightgbm` relies on them.

HOWEVER... their brief presence in that list tells `mypy` "well sometimes there can be `None` in this list". And honestly, it confused me too.

This PR proposes making `self.best_score_list` easier for both `mypy` and humans to understand by removing that `None`-padding, and instead changing the update logic to *"if it's the first time this list is being updated, build it up by `.append()`-ing every element of `env.evaluation_result_list` to it"*.